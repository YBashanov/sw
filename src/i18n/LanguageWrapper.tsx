import React, { createContext, useState } from 'react';
import { getIntlLocale, getIntlMessages } from '@/i18n';
import { IntlProvider } from 'react-intl';

export type Language = 'ru' | 'en';

export const defaultLanguage: Language = 'ru';

type LanguageContextType = {
    language?: Language;
    setLanguage?: (language: Language) => void;
};

type LanguageWrapperType = {
    children: React.ReactNode;
};

export const LanguageContext = createContext<LanguageContextType>({
    language: defaultLanguage,
});

export const defaultSaveLang = localStorage.getItem('language') as Language;

const LanguageWrapper: React.FC<LanguageWrapperType> = ({ children }) => {
    const [language, setLanguage] = useState<Language>(defaultSaveLang || defaultLanguage);

    const handleSetLanguage = (lang: Language) => {
        localStorage.setItem('language', lang);

        // установка для React-компонентов
        setLanguage(lang);

        // установка для статики (схемы Yup, модели Redux)
        // устанавливается в i18n/index

        // требуется перезагрузка для сохранения в localStorage и взятия оттуда
        location.href = '/';
    };

    // тестирование смены языка. (в идеале это должно быть по кнопке)
    React.useEffect(() => {
        // const wantLang = 'ru';
        // if (defaultSaveLang !== wantLang) handleSetLanguage(wantLang);
    }, []);

    return (
        <LanguageContext.Provider value={{ language, setLanguage: handleSetLanguage }}>
            <IntlProvider locale={getIntlLocale(language)} messages={getIntlMessages(language)}>
                {children}
            </IntlProvider>
        </LanguageContext.Provider>
    );
};

export default LanguageWrapper;
