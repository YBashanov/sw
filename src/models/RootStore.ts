import { configureStore } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import thunk from 'redux-thunk';
import { reducers } from './reducers';

/**
 * Главное хранилище приложения
 */
const RootStore = configureStore({
    reducer: reducers,
    middleware: [thunk] as const,
    devTools: process.env.NODE_ENV !== 'production',
});

export type RootState = ReturnType<typeof RootStore.getState>;

export type AppDispatch = typeof RootStore.dispatch;
export const useAppDispatch: () => AppDispatch = useDispatch;

export default RootStore;
