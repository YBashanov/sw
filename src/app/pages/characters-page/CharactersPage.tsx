import React from 'react';
import { Helmet } from 'react-helmet';
import { PREFIX_CLS } from '@/constants';
import { AnyAction } from '@reduxjs/toolkit';
import PageLayout from '@/app/pages/-page-layout';
import { useDispatch, useSelector } from 'react-redux';
import { Table, Spin } from 'antd';
import { PaginationProps } from 'antd/lib/pagination';
import { tableColumns } from './tableColumns';
import { Actions, ActionsSync, Selectors, Convertors } from '@/models/characters';
import { useIntl } from 'react-intl';
import { CharactersFilters } from '@/app/pages/characters-page';

import './index.less';

const mainClassName = `${PREFIX_CLS}-characters-page`;

/**
 * Список всех персонажей
 */
const CharactersPage: React.FC = () => {
    const { formatMessage: f } = useIntl();
    const dispatch = useDispatch();
    const pageState = useSelector(Selectors.getState);

    React.useEffect(() => {
        const request = Convertors.requestGetTable(pageState);
        const result = dispatch(
            ActionsSync.getTableData({
                ...request,
                hasLocal: '',
            })
        );

        if (result.payload.hasLocal === '0') {
            dispatch(Actions.getTableData(request) as unknown as AnyAction);
        }
    }, [dispatch, pageState.pagination.current, pageState.search, pageState.filters]);

    /**
     * Переход по страницам (пагинация)
     */
    const onChange = (pagination: PaginationProps) => {
        dispatch(
            ActionsSync.setPagination({
                ...pageState.pagination,
                current: pagination.current,
            })
        );
    };

    return (
        <div className={mainClassName}>
            <Helmet>
                <title>{f({ id: 'App.Title.Characters' })}</title>
            </Helmet>
            <CharactersFilters />
            <Table
                dataSource={pageState.tableData}
                rowKey="url"
                columns={tableColumns}
                pagination={pageState.pagination}
                onChange={onChange}
                loading={{ indicator: <Spin />, spinning: !pageState.isLoad }}
            />
        </div>
    );
};

export default PageLayout(CharactersPage);
