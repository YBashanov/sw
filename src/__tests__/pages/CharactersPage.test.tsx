import React from 'react';
import { describe, expect, it } from '@jest/globals';
import { render } from 'enzyme';
import CharactersPage from '@/app/pages/characters-page/CharactersPage';
import LanguageWrapper from '@/i18n/LanguageWrapper';
import RootStore from '@/models/RootStore';
import { Provider } from 'react-redux';

describe('testing: CharactersPage', function () {
    const JSX = (
        <Provider store={RootStore}>
            <LanguageWrapper>
                <CharactersPage />
            </LanguageWrapper>
        </Provider>
    );
    const Component = render(JSX);

    it('Snapshot correctly', function () {
        expect(Component).toMatchSnapshot();
    });

    it('Static HTML', function () {
        expect(Component.text()).toEqual(
            'ФильтрыСозданИмяГод рожденияЦвет глазЦвет волосЦвет кожиПолВысотаМассаФильмыNo dataДополнительноВыбор темыСтандартРозовыйЗеленыйВыбор языкаRUEN'
        );
    });
});
