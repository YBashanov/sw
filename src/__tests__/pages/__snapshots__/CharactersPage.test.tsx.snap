// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing: CharactersPage Snapshot correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="sw-page-layout"
  >
    <div
      class="sw-page-layout-content"
    >
      <div
        class="sw-page-layout-content-page"
      >
        <div
          class="sw-characters-page"
        >
          <div
            class="sw-characters-page-filters"
          >
            <div
              class="ant-space css-dev-only-do-not-override-ed5zg0 ant-space-horizontal ant-space-align-center"
            >
              <div
                class="ant-space-item"
                style="margin-right:8px"
              >
                <span
                  class="ant-input-group-wrapper ant-input-search ant-input-search-with-button css-dev-only-do-not-override-ed5zg0"
                >
                  <span
                    class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-ed5zg0"
                  >
                    <input
                      class="ant-input css-dev-only-do-not-override-ed5zg0"
                      placeholder="Поиск по имени"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-group-addon"
                    >
                      <button
                        class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary ant-input-search-button"
                        type="button"
                      >
                        <span
                          aria-label="search"
                          class="anticon anticon-search"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="search"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </span>
              </div>
              <div
                class="ant-space-item"
                style="margin-right:8px"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
                  type="button"
                >
                  <span>
                    Фильтры
                  </span>
                </button>
              </div>
              <div
                class="ant-space-item"
              />
            </div>
          </div>
          <div
            class="ant-table-wrapper css-dev-only-do-not-override-ed5zg0"
          >
            <div
              class="ant-spin-nested-loading css-dev-only-do-not-override-ed5zg0"
            >
              <div
                class="ant-spin-container"
              >
                <div
                  class="ant-table ant-table-empty"
                >
                  <div
                    class="ant-table-container"
                  >
                    <div
                      class="ant-table-content"
                    >
                      <table
                        style="table-layout:auto"
                      >
                        <colgroup />
                        <thead
                          class="ant-table-thead"
                        >
                          <tr>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Создан
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Имя
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Год рождения
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Цвет глаз
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Цвет волос
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Цвет кожи
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Пол
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Высота
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Масса
                            </th>
                            <th
                              class="ant-table-cell"
                              scope="col"
                            >
                              Фильмы
                            </th>
                          </tr>
                        </thead>
                        <tbody
                          class="ant-table-tbody"
                        >
                          <tr
                            class="ant-table-placeholder"
                          >
                            <td
                              class="ant-table-cell"
                              colspan="10"
                            >
                              <div
                                class="css-dev-only-do-not-override-ed5zg0 ant-empty ant-empty-normal"
                              >
                                <div
                                  class="ant-empty-image"
                                >
                                  <svg
                                    height="41"
                                    viewBox="0 0 64 41"
                                    width="64"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <g
                                      fill="none"
                                      fill-rule="evenodd"
                                      transform="translate(0 1)"
                                    >
                                      <ellipse
                                        cx="32"
                                        cy="33"
                                        fill="#f5f5f5"
                                        rx="32"
                                        ry="7"
                                      />
                                      <g
                                        fill-rule="nonzero"
                                        stroke="#d9d9d9"
                                      >
                                        <path
                                          d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                        />
                                        <path
                                          d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                          fill="#fafafa"
                                        />
                                      </g>
                                    </g>
                                  </svg>
                                </div>
                                <div
                                  class="ant-empty-description"
                                >
                                  No data
                                </div>
                              </div>
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sw-page-layout-footer"
    >
      <div
        class="separator"
      >
        <div
          class="label"
        >
          Дополнительно
        </div>
      </div>
      <div
        class="theme-controls"
      >
        <div
          class="label"
        >
          Выбор темы
        </div>
        <div
          class="controls"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
            type="button"
          >
            <span>
              Стандарт
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            type="button"
          >
            <span>
              Розовый
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            type="button"
          >
            <span>
              Зеленый
            </span>
          </button>
        </div>
      </div>
      <div
        class="language-controls"
      >
        <div
          class="label"
        >
          Выбор языка
        </div>
        <div
          class="controls"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
            id="ru"
            type="button"
          >
            <span>
              RU
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            id="en"
            type="button"
          >
            <span>
              EN
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
